CAS.o: Add some documentation for CASObjectFormats

dexonsmith opened the pull request:
Just a start; hopefully enough to be useful though. Expectation is that we keep this up-to-date as we make changes.

dexonsmith committed:
CAS.o: Add some documentation for CASObjectFormats

dexonsmith commented:
I don't think this hooks up sphinx to actually build the docs; for now they're just markdown documents available in-tree.

@cachemeifyoucan , if you'd like to document FlatV1 differently / more detail / etc., I'm thinking you can follow up in tree. WDYT?

cachemeifyoucan commented on llvm/docs/CASObjectFormats/CASObjectFormats.md:
```diff
@@ -0,0 +1,78 @@
+
+Not clear whether we'll evolve one of these or create a new one.
+
+### Caveat: Does not support real algorithms
```
This feels more like just we don't have an interface that compiler or linker can call into.
LinkGraph totally supports linking concepts and FlatV1 can totally do a fair amount of lazy loading if needed. We just don't have API for those.

dexonsmith replied:
Rewored to "Not ready to support" and clarified text a bit.

cachemeifyoucan commented on llvm/docs/CASObjectFormats/FlatV1.md:
```diff
@@ -0,0 +1,83 @@
+concepts outlined to reduce aggregate storage cost.
+
+**compile-unit**: all graph information stored at the top-level
+- Data: all symbols, blocks, sections, referring to templates with a 4B index
```
Compile-unit in flatv1 is basically:
* Template array: CASIDs to other CAS objects that stores Symbol, Blocks, etc.
* Data: an array of indexes that encode all the references from one entities to another. The array works based on the fact that there is a deterministic traversal order to visit all the sections, symbols and blocks and each of them know how many indexes they need when visiting.

dexonsmith replied:
Thanks; cleaned up the description a bit!

cachemeifyoucan reviewed (approved):
I think this is good and we can improve in tree.

dexonsmith committed:
CAS.o: Fine-tune docs for CASObjectFormats after review

After some review:

- Improve description of FlatV1
- Clarify "not ready for real algorithms" caveat
- Also add newlines before lists to support a better diversity of
  markdown viewers
